crates/weaver_common/src/result.rs

```diff
@@ -0,0 +1,230 @@
+// SPDX-License-Identifier: Apache-2.0
+
+//! Weaver Result type supporting both non-fatal errors (NFEs) and fatal errors.
+//!
+//! NFEs do not prevent the next operations from completing successfully. For example,
+//! if a semconv file is invalid, we generate a non-fatal error and continue processing
+//! the other files.
+//!
+//! NFEs in Weaver are standard Rust errors.
+
+use crate::diagnostic::{DiagnosticMessage, DiagnosticMessages};
+use crate::error::WeaverError;
+use miette::Diagnostic;
+use serde::Serialize;
+use std::error::Error;
+
+/// Weaver Result type supporting both non-fatal errors (NFEs) and fatal errors.
+#[must_use]
+pub enum WResult<T, E> {
+    /// The operation was successful, the result T is returned.
+    Ok(T),
+    /// The operation was successful, the result T is returned along with
+    /// a list of non-fatal errors.
+    OkWithNFEs(T, Vec<E>),
+    /// The operation failed with a fatal error. By definition, we can only have
+    /// one fatal error.
+    FatalErr(E),
+}
+
+impl<T, E> WResult<T, E>
+where
+    E: WeaverError<E> + Error + Diagnostic + Serialize + Send + Sync + 'static,
+{
+    /// Creates a new [`WResult`] with a successful result.
+    pub fn with_non_fatal_errors(result: T, non_fatal_errors: Vec<E>) -> Self {
+        if non_fatal_errors.is_empty() {
+            WResult::Ok(result)
+        } else {
+            WResult::OkWithNFEs(result, non_fatal_errors)
+        }
+    }
+
+    /// Converts a [`WResult`] into a standard [`Result`], optionally capturing non-fatal errors.
+    pub fn capture_non_fatal_errors(self, diag_msgs: &mut DiagnosticMessages) -> Result<T, E> {
+        match self {
+            WResult::Ok(result) => Ok(result),
+            WResult::OkWithNFEs(result, nfes) => {
+                let msgs = nfes.into_iter().map(DiagnosticMessage::new).collect();
+                diag_msgs.extend_from_vec(msgs);
+                Ok(result)
+            }
+            WResult::FatalErr(fatal_err) => Err(fatal_err),
+        }
+    }
+
+    /// Capture the warnings into the provided vector and return a [`WResult`]
+    /// without the warnings.
+    pub fn capture_warnings(self, diag_msgs: &mut DiagnosticMessages) -> WResult<T, E> {
+        if let WResult::OkWithNFEs(result, nfes) = self {
+            let (warnings, errors): (Vec<_>, Vec<_>) = nfes
+                .into_iter()
+                .partition(|e| matches!(e.severity(), Some(miette::Severity::Warning)));
+            let warnings: Vec<_> = warnings.into_iter().map(DiagnosticMessage::new).collect();
+            diag_msgs.extend_from_vec(warnings);
+            if errors.is_empty() {
+                WResult::Ok(result)
+            } else {
+                WResult::OkWithNFEs(result, errors)
+            }
+        } else {
+            self
+        }
+    }
+
+    /// Return a [`WResult`] without the non-fatal errors with severity=warning.
+    pub fn ignore_severity_warnings(self) -> WResult<T, E> {
```

Comment:
This is fine for now, but we should consider an `ignore(F) -> WResult<T,E> where F: FnOnce(E) -> Bool`

Reply:
That’s something that crossed my mind at some point, but then I forgot about it. I will change that before the merge.